File: ltkit/cpp/tests/smoke.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Michael King — part of ltkit (https://github.com/BTFProdigy/ltkit)
// Licensed under the Apache License, Version 2.0. See the LICENSE and NOTICE
// files in the project root. Attribution must be retained in derivative works.
// Engine smoke test over a pure in-memory mock model — no tensor backend.
// Mirrors the Rust/Python invariant: sparsity rises monotonically and the
// final masks agree with the model's zeroed weights (mask-persistence).
#include "ltkit/ltkit.hpp"

#include <cassert>
#include <cstdio>
#include <map>
#include <string>
#include <unordered_map>
#include <vector>

using ltkit::Criterion;
using ltkit::ImpConfig;
using ltkit::RewindPolicy;
using ltkit::Scope;

struct MockModel {
    using State = std::unordered_map<std::string, std::vector<float>>;

    std::unordered_map<std::string, std::vector<float>> weights;
    std::unordered_map<std::string, std::vector<bool>> masks;

    MockModel() {
        std::vector<float> a, b;
        for (int i = 1; i <= 20; ++i) a.push_back(static_cast<float>(i));
        for (int i = 1; i <= 12; ++i) b.push_back(static_cast<float>(i) * 0.5f);
        weights["a.weight"] = a;
        weights["b.weight"] = b;
        for (auto& kv : weights) masks[kv.first] = std::vector<bool>(kv.second.size(), true);
    }

    void enforce() {
        for (auto& kv : weights) {
            auto& m = masks[kv.first];
            for (std::size_t i = 0; i < kv.second.size(); ++i)
                if (!m[i]) kv.second[i] = 0.0f;
        }
    }

    std::vector<std::string> parameter_groups() const {
        std::vector<std::string> g;
        for (auto& kv : weights) g.push_back(kv.first);
        std::sort(g.begin(), g.end());
        return g;
    }
    std::vector<float> scores(const std::string& name, Criterion) const {
        std::vector<float> s;
        for (float w : weights.at(name)) s.push_back(w < 0 ? -w : w);
        return s;
    }
    void apply_mask(const std::string& name, const std::vector<bool>& mask) {
        masks[name] = mask;
        enforce();
    }
    State snapshot() const { return weights; }
    void restore(const State& s) {
        weights = s;
        enforce();  // masks survive restore
    }
    void fit(std::size_t) { enforce(); }  // training must not revive pruned weights
    float evaluate() const {
        float total = 0;
        for (auto& kv : weights)
            for (float w : kv.second) total += (w < 0 ? -w : w);
        return total / 100.0f;
    }
};

int main() {
    MockModel model;
    ImpConfig cfg;
    cfg.rounds = 4;
    cfg.prune_rate = 0.3f;
    cfg.epochs_per = 1;
    cfg.scope = Scope::Global;
    cfg.rewind = RewindPolicy::Init;

    auto res = ltkit::run_imp(model, cfg);

    assert(res.history.front().sparsity == 0.0f);
    for (std::size_t i = 1; i < res.history.size(); ++i)
        assert(res.history[i].sparsity >= res.history[i - 1].sparsity);
    assert(res.history.back().sparsity > 0.5f);

    for (auto& kv : res.masks) {
        const auto& w = model.weights[kv.first];
        for (std::size_t i = 0; i < kv.second.size(); ++i)
            if (!kv.second[i]) assert(w[i] == 0.0f);
    }
    std::printf("cpp smoke OK  final_sparsity=%.3f\n", res.history.back().sparsity);
    return 0;
}

File: ltkit/rust/tests/smoke.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Michael King — part of ltkit (https://github.com/BTFProdigy/ltkit)
// Licensed under the Apache License, Version 2.0. See the LICENSE and NOTICE
// files in the project root. Attribution must be retained in derivative works.
//! Engine smoke test over a pure in-memory mock model — no tensor backend.
//!
//! Mirrors the Python `test_smoke.py` invariant: sparsity rises monotonically,
//! the final returned masks agree with the model's zeroed weights, and a pruned
//! element stays zero across `fit`/`restore` (mask-persistence invariant).

use std::collections::HashMap;

use ltkit::contract::{Criterion, PrunableModel, RewindPolicy};
use ltkit::imp::{run_imp, ImpConfig, Scope};

struct MockModel {
    weights: HashMap<String, Vec<f32>>,
    masks: HashMap<String, Vec<bool>>,
}

impl MockModel {
    fn new() -> Self {
        let mut weights: HashMap<String, Vec<f32>> = HashMap::new();
        // deterministic, distinct magnitudes so pruning order is well-defined
        weights.insert("a.weight".into(), (1..=20).map(|i| i as f32).collect());
        weights.insert("b.weight".into(), (1..=12).map(|i| (i as f32) * 0.5).collect());
        let masks = weights
            .iter()
            .map(|(k, v)| (k.clone(), vec![true; v.len()]))
            .collect();
        MockModel { weights, masks }
    }

    fn enforce(&mut self) {
        for (name, w) in self.weights.iter_mut() {
            if let Some(m) = self.masks.get(name) {
                for (wi, &keep) in w.iter_mut().zip(m.iter()) {
                    if !keep {
                        *wi = 0.0;
                    }
                }
            }
        }
    }
}

impl PrunableModel for MockModel {
    type State = HashMap<String, Vec<f32>>;

    fn parameter_groups(&self) -> Vec<String> {
        let mut g: Vec<String> = self.weights.keys().cloned().collect();
        g.sort();
        g
    }

    fn scores(&self, name: &str, _c: Criterion) -> Vec<f32> {
        self.weights[name].iter().map(|w| w.abs()).collect()
    }

    fn apply_mask(&mut self, name: &str, mask: &[bool]) {
        self.masks.insert(name.to_string(), mask.to_vec());
        self.enforce();
    }

    fn snapshot(&self) -> Self::State {
        self.weights.clone()
    }

    fn restore(&mut self, state: &Self::State) {
        self.weights = state.clone();
        self.enforce(); // masks survive restore
    }

    fn fit(&mut self, _epochs: usize) {
        self.enforce(); // training must not revive pruned weights
    }

    fn evaluate(&self) -> f32 {
        // fraction of surviving weight mass — purely to have a metric
        let total: f32 = self.weights.values().flatten().map(|w| w.abs()).sum();
        total / 100.0
    }
}

#[test]
fn imp_engine_invariants() {
    let mut model = MockModel::new();
    let cfg = ImpConfig {
        rounds: 4,
        prune_rate: 0.3,
        epochs_per: 1,
        criterion: Criterion::Magnitude,
        rewind: RewindPolicy::Init,
        scope: Scope::Global,
        early_k_epochs: 0,
        verbose: false,
    };
    let res = run_imp(&mut model, &cfg);

    // sparsity rises monotonically, starts at 0
    let sp: Vec<f32> = res.history.iter().map(|r| r.sparsity).collect();
    assert_eq!(sp[0], 0.0, "{sp:?}");
    for w in sp.windows(2) {
        assert!(w[1] >= w[0], "non-monotonic sparsity {sp:?}");
    }
    assert!(*sp.last().unwrap() > 0.5, "final sparsity too low {sp:?}");

    // final masks agree with zeroed weights (persistence across whole run)
    for (name, mask) in res.masks.iter() {
        let w = &model.weights[name];
        for (i, &keep) in mask.iter().enumerate() {
            if !keep {
                assert_eq!(w[i], 0.0, "pruned weight {name}[{i}] not zero");
            }
        }
    }
}